sky/packages/sky/lib/src/widgets/page.dart

```diff
@@ -0,0 +1,137 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/animation.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'navigator2.dart';
+import 'overlay.dart';
+import 'transitions.dart';
+
+abstract class TransitionRoute extends Route {
+  bool get opaque => true;
+
+  PerformanceView get performance => _performance?.view;
+  Performance _performance;
+
+  Duration get transitionDuration;
+
+  Performance createPerformance() {
+    Duration duration = transitionDuration;
+    assert(duration != null && duration >= Duration.ZERO);
+    return new Performance(duration: duration, debugLabel: debugLabel);
+  }
+
+  dynamic _result;
+
+  void _handleStatusChanged(PerformanceStatus status) {
+    if (status == PerformanceStatus.completed && opaque) {
+      bottomEntry.opaque = true;
+    } else if (status == PerformanceStatus.dismissed) {
+      super.remove(_result);
+    }
+  }
+
+  void add(OverlayState overlayer, OverlayEntry insertionPoint) {
+    _performance = createPerformance()
+      ..addStatusListener(_handleStatusChanged)
+      ..forward();
+    super.add(overlayer, insertionPoint);
+  }
+
+  void remove(dynamic result) {
+    _result = result;
+    _performance.reverse();
+  }
+
+  String get debugLabel => '$runtimeType';
+  String toString() => '$runtimeType(performance: $_performance)';
+}
+
+class _Page extends StatefulComponent {
+  _Page({
+    Key key,
+    PageRoute route
+  }) : route = route, super(key: key);
+
+  final PageRoute route;
+
+  _PageState createState() => new _PageState();
+}
+
+class _PageState extends State<_Page> {
+  final AnimatedValue<Point> _position =
+      new AnimatedValue<Point>(const Point(0.0, 75.0), end: Point.origin, curve: Curves.easeOut);
+
+  final AnimatedValue<double> _opacity =
+      new AnimatedValue<double>(0.0, end: 1.0, curve: Curves.easeOut);
+
+  final GlobalKey _subtreeKey = new GlobalKey();
+
+  Widget build(BuildContext context) {
+    if (config.route._offstage) {
+      return new OffStage(
+        child: new KeyedSubtree(
+          key: _subtreeKey,
+          child: _invokeBuilder()
+        )
+      );
+    }
+    return new SlideTransition(
```

Comment:
Is there a way for the Page to know whether it caused any heroes to be in flight, and if so, to not bother with the slide transition and fade transition? (Consider how wipe transitions will work.)